syntax = "proto3";

package fennel.proto;

import "schema.proto";
import "status.proto";

enum FeatureDefType {
  FEATURE = 0;
  FEATURE_PACK = 1;
}

message CreateFeatureRequest {
  string name = 1;
  string mode = 2;
  FeatureDefType type = 3;
  uint32 version = 4;
  bytes function = 5;
  string function_source_code = 6;
  repeated string depends_on_aggregates = 7;
  repeated string depends_on_features = 8;
  Schema schema = 9;
  string workspace = 10;
}

message FeatureRef {
  string name = 1;
  string workspace = 2;
}

message FeatureValue {
  FeatureRef feature = 1;
  bytes value = 2;
}

message ExtractFeaturesRequest {
  repeated FeatureRef features = 1;
  bytes args = 2;
  map<string, float> config = 3;
}

message CreateFeatureResponse { Status status = 1; }

message ExtractFeaturesResponse {
  Status status = 1;
  FeatureValue value = 2;
}
